### PR TITLE
libp2p helper updates

### PR DESCRIPTION
**Pull request by noot**
Hey guys! I've completed all the specified libp2p helper updates. Please let me know your feedback :)

Explain your changes here.

- Update openStreamMsg.run to not call handleStreamReads
- Update handleStreamReads to close stream before returning
- Update closeStreamMsg to delete stream from app.Streams mapping (need clarification)

- Update addPeerMsg.run to directly connect to peer instead of using bootstrap.Bootstrap
- Update beginAdvertisingMsg.run to connect directly to bootstrap peers, and use dht.Bootstrap instead of using - bootstrap.Bootstrap 

- Update beginAdvertisingMsg.run to use routingDiscovery.FindPeers to find peers and call foundPeer, instead of using EventBus
- Test that peers can be discovered using DHT and mDNS
- added `NoMDNS` and `NoDHT` bools to `app` for testing this

- Use go-mplex library instead of fork and change MaxMessageSize 1 << 30
- Test to make sure message of desired size can still be sent

- Update libp2p.New to use BandwidthReporter option
- Store BandwidthReporter inside codanet.Helper for access

- Update app.Validators[seqno] setting in pubsub validation function

- Start goroutine to regularly check bandwidth used by its connected peers using BandwidthReporter.GetBandwidthByPeer
- Start goroutine to regularly record latency using Peerstore.RecordLatency
- In same goroutine, regularly calculate avg of latency using Peerstore.LatencyEWMA
- Expose bandwidth for each peer and latency for each peer using Prometheus

- run `golangci-lint` and fix all lint errors
- run `go mod tidy` to cleanup go.mod and go.sum file

Explain how you tested your changes here.

in project root:
```
cd src/app/libp2p_helper/src
go test ./...
```

Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

File: src/app/libp2p_helper/src/libp2p_helper/main.go [hunk 1]
```diff
@@ -919,20 +911,26 @@ func (ap *beginAdvertisingMsg) run(app *app) (interface{}, error) {
...
 		}
 	}
 
-	if len(app.AddedPeers) > 0 {
-		app.Bootstrapper, err = bootstrap.Bootstrap(app.P2p.Me, app.P2p.Host, app.P2p.Dht, bootstrap.BootstrapConfigWithPeers(app.AddedPeers))
+	for _, info := range app.AddedPeers {
+		err = app.P2p.Host.Connect(context.Background(), info)
 		if err != nil {
-			return nil, badp2p(err)
+			// perhaps some logging
```
Comment (bkase) [inline src/app/libp2p_helper/src/libp2p_helper/main.go:928]: could you add a log here?

**Review by bkase — COMMENTED** [review]
Otherwise everything is looking good!

File: src/app/libp2p_helper/src/libp2p_helper/main.go (hunk 1 above)
Comment (noot) [reply src/app/libp2p_helper/src/libp2p_helper/main.go]: done!

**Review by emberian — APPROVED** [review]
these changes look great!

**Comment by emberian** [comment]
@MinaProtocol/infra-eng-reviewers after this PR, the helper runs a prometheus server exposing latency/bandwidth metrics. this probably needs some infra work to integrate?

**Comment by bkase** [comment]
@nholland94 and I talked with @noot and she's going to make sure to (at least) run-local-network.sh with mDNS on and off before we land this. @lk86 is interested in trying these changes on a node for the public network too

**Comment by noot** [comment]
@bkase just ran `run_local_network.sh` with both `app.NoMDNS = false` and `app.NoMDNS = true`, in both cases I'm able to connect to the other 3 peers, eg for `fish`: 

```
$ _build/default/src/app/cli/src/coda.exe client status -daemon-port 5005
{"timestamp":"2020-12-09 00:25:31.687514Z","level":"Debug","source":{"module":"Block_time","location":"File \"src/lib/block_time/block_time.ml\", line 76, characters 18-30"},"message":"Environment variable CODA_TIME_OFFSET not found, using default of 0","metadata":{"pid":27335}}
Coda daemon status
-----------------------------------

Global number of accounts:       246
Block height:                    1
Max observed block length:       37
Local uptime:                    3m10s
Ledger Merkle root:              jwqrqt64qKCZuGaD168ZvTBVfS7xKSDEL1xxRFRb8duzA9gCiWy
Protocol state hash:             3NKJHQiJrkmvMq52AeeT97BdK5uTcW2ipyKCwVugy3xwYp7bhF5o
Chain id:                        b6d84ac9737187eaedfe2b22a72d3ec29521e973c884c0e99ae7291400e8ec39
Git SHA-1:                       cb3736a77f2137ba5239b7159e10b7275c62cf89
Configuration directory:         /tmp/mina-local-network-1-1-1/nodes/fish_1
Peers:                           3 (127.0.0.1:4007 192.168.0.103:6007 127.0.0.1:3002)
User_commands sent:              0
SNARK worker:                    B62qp4UturELw4MmhAZhor8rwzaH1BBAivRnvdp1Yhkq6odhhFiT8uC
SNARK work fee:                  1000000000
Sync status:                     Catchup
Block producers running:         1 (B62qrQr5TWN1FVcNMTxkKmm46MDCj13gqdyFiLgRuGySWaRNRsNepxi)
Best tip consensus time:         epoch=0, slot=0
Next block will be produced in:  None this epoch… checking at in 15.981m
Consensus time now:              epoch=0, slot=96
Consensus mechanism:             proof_of_stake
Consensus configuration:         
	Delta:                     0
	k:                         24
	Slots per epoch:           576
	Slot duration:             2s
	Epoch duration:            19m12s
	Genesis state timestamp:   2020-12-09 00:22:18.538000Z
	Acceptable network delay:  2s

Addresses and ports:             
	External IP:    38.88.101.233
	Bind IP:        0.0.0.0
	Libp2p PeerID:  12D3KooWRUUPF1jz65Sic6G1nJuUMYVAWDo7Fue2E4y5vUCFfrR3
	Libp2p port:    5007
	Client port:    5005
```

**Review by bkase — APPROVED** [review]

**Comment by bkase** [comment]
!ci-build-me

**Comment by bkase** [comment]
I rolled this up into https://github.com/MinaProtocol/mina/pull/6992

**Pull request closed without merging**